File: src/Year2018/day03.cpp
#include <bits/stdc++.h>

struct Rect {
    int top, left, height, width;
};

using Grid = std::vector<std::vector<int>>;

bool overlap(const Rect& r1, const Rect& r2) {
    if (r1.left > r2.left) {
        return overlap(r2, r1);
    }

    return r1.left + r1.width > r2.left &&
           r1.top + r1.height > r2.top &&
           r2.top + r2.height > r1.top;
}

int main() {
    std::vector<Rect> rects;

    int maxwidth = 0, maxheight = 0;

    char c;
    while (std::cin >> c) {
        Rect r;
        std::string s;
        std::cin >> s >> s
                 >> r.top
                 >> c
                 >> r.left
                 >> c
                 >> r.height
                 >> c
                 >> r.width;

        r.top++;
        r.left++;

        rects.push_back(r);

        maxwidth  = std::max(r.left + r.width , maxwidth);
        maxheight = std::max(r.top  + r.height, maxheight);
    }

    const int Width = maxwidth + 1;
    const int Height = maxheight + 1;

    Grid grid (Width, std::vector<int> (Height, 0));

    for (const auto& r : rects) {
        grid[r.left][r.top]++;
        grid[r.left][r.top + r.height]--;
        grid[r.left + r.width][r.top]--;
        grid[r.left + r.width][r.top + r.height]++;
    }

    int doubly_claimed = 0;
    for (int x = 1; x < Width; ++x) {
        for (int y = 1; y < Height; ++y) {
            grid[x][y] += grid[x][y-1] + grid[x-1][y] - grid[x-1][y-1];

            if (grid[x][y] > 1) {
                doubly_claimed ++;
            }
        }
    }

    std::cout << "Solution Part One: " << doubly_claimed << '\n';

    std::vector<bool> ok (rects.size(), true);

    for (std::size_t i = 0; i < rects.size(); ++i) {
        for (std::size_t j = i + 1; j < rects.size(); ++j) {
            if (!ok[i] && !ok[j]) continue;

            if (overlap(rects[i], rects[j])) {
                ok[i] = false;
                ok[j] = false;
            }
        }

        if (ok[i]) {
            std::cout << "Solution Part Two: " << i + 1 << '\n'; // < 1094
            break;
        }
    }
}

File: src/Year2018/Day03.hs
module Year2018.Day03 (solveA, solveB) where

import Data.Void (Void)
import Text.Megaparsec (Parsec, parse, sepEndBy1, eof)
import Text.Megaparsec.Char
import qualified Text.Megaparsec.Char.Lexer as L
import Text.Megaparsec.Error

import PrefixSumGrid (PrefixSumGrid)
import qualified PrefixSumGrid as PSG
import Data.List (foldl', sort)

import Debug.Trace

type Parser = Parsec Void String

number :: Parser Int
number = L.decimal

lexeme :: Parser a -> Parser a
lexeme = L.lexeme space1

data Rectangle = Rectangle { left :: Int
                           , top :: Int
                           , width :: Int
                           , height :: Int
                           } deriving (Show)

parseLine :: Parser (Int, Rectangle)
parseLine = do
    char '#'
    i <- number
    space1
    char '@'
    space1
    l <- number
    char ','
    t <- number
    char ':'
    space1
    w <- number
    char 'x'
    h <- number

    return (i, Rectangle l t w h)

readInput :: String -> [(Int, Rectangle)]
readInput input =
    case parse (sepEndBy1 parseLine newline <* eof) "problem input" input of
        Left err  -> error $ parseErrorPretty err
        Right res -> res

intersect :: Rectangle -> Rectangle -> Bool
intersect r1 r2 =
    if left r1 > left r2 then
        intersect r2 r1
    else
        left r1 + width r1 > left r2  &&
        top r1 + height r1 > top r2 &&
        top r2 + height r2 > top r1

gridSize :: [Rectangle] -> (Int, Int)
gridSize [] = (0, 0)
gridSize (Rectangle l t w h : rest) =
    (max (l + w) rx, max (t + h) ry)
    where
        (rx, ry) = gridSize rest

solveA :: String -> Int
solveA input = sum $ map (doubleClaimLength 0 0 . events) [0..maxy-1]
    where
        rects = map snd $ readInput input
        (_,maxy) = gridSize rects
        events y = sort $ do
            Rectangle l t w h <- rects
            if y >= t && y < t + h then
                [(l, 1), (l + w, -1)]
            else
                []
        doubleClaimLength _ _ [] = 0
        doubleClaimLength level lastPos ((pos, diff):xs) =
            doubleClaimLength (level + diff) pos xs + added
            where
                added = if level >= 2 then pos - lastPos else 0


solveB :: String -> Int
solveB input = fst . head $ filter good rects
    where
        rects = readInput input
        except n = filter (\(i, r) -> i /= n)
        good (idx, rect) = all (not . intersect rect) . map snd $ except idx rects